generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -691,6 +697,105 @@ public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
 
   }
 
+  /**
+   * Identity transformation for integer types (both unsigned and signed, and their vectors) that
+   * double bitwise inverts an integer, producing the same integer as output. When performed,
+   * transforms an expression, e, such that:
+   *    e -> ~(~(expr)).
+   */
+  private class IdentityBitwiseNotNot extends AbstractIdentityTransformation {
+    private IdentityBitwiseNotNot() {
+      super(BasicType.allIntegerTypes(), false);
+    }
+
+    @Override
+    public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
+                      Fuzzer fuzzer) {
+      assert BasicType.allIntegerTypes().contains(type);
+      // Invert once
+      Expr result = new UnaryExpr(new ParenExpr(applyIdentityFunction(expr, type,
+          constContext,
+          depth, fuzzer)), UnOp.BNEG);
+      // Invert again
+      result = new UnaryExpr(new ParenExpr(applyIdentityFunction(result, type, constContext,
+          depth, fuzzer)), UnOp.BNEG);
+      return identityConstructor(expr, result);
+    }
+  }
+
+  /**
+   * Identity transformation for integer types (both unsigned and signed, and their vectors) that
+   * ORs an integer with itself, producing the same integer as output. When performed, transforms an
+   * expression, e, such that:
+   *    e -> (e) | (e).
+   */
+  private class IdentityBitwiseOr extends AbstractIdentityTransformation {
+    private IdentityBitwiseOr() {
```

Comment:
The second argument to "super" should be "true" here, since we're going to end up evaluating the mutated expression twice.  That means that it (in general) needs to be side-effect free.

Reply:
It would be worth adding a note too that effect in the comment for the private class.

Reply:
Done. Also see my comments in the below conversation about the comment, as they're related somewhat to this.